addons/l10n_it_edi_doi/models/sale_order.py

```diff
@@ -0,0 +1,266 @@
+                continue
+
+            company = order.company_id
+            partner = order.partner_id.commercial_partner_id
+            date = order.l10n_it_edi_doi_date
+            currency = order.currency_id
+
+            # Avoid a query or changing a manually set declaration of intent
+            # (if the declaration is still valid).
+            current_declaration = order.l10n_it_edi_doi_id
+            if current_declaration and not current_declaration._get_validity_warnings(company, partner, currency, date, sales_order=True):
+                continue
+
+            declaration = self.env['l10n_it_edi_doi.declaration_of_intent']\
+                ._fetch_valid_declaration_of_intent(company, partner, currency, date)
+            order.l10n_it_edi_doi_id = declaration
+
+    @api.depends('tax_totals', 'invoice_ids', 'invoice_ids.state', 'invoice_ids.tax_totals')
+    def _compute_l10n_it_edi_doi_not_yet_invoiced(self):
+        for order in self:
+            declaration = order.l10n_it_edi_doi_id
+            order.l10n_it_edi_doi_not_yet_invoiced = order._l10n_it_edi_doi_get_amount_not_yet_invoiced(declaration)
+
+    @api.depends('l10n_it_edi_doi_id', 'l10n_it_edi_doi_id.remaining', 'state', 'tax_totals')
+    def _compute_l10n_it_edi_doi_warning(self):
+        for order in self:
+            order.l10n_it_edi_doi_warning = ''
+            declaration = order.l10n_it_edi_doi_id
+
+            show_warning = declaration and order.state != 'cancelled'
+            if not show_warning:
+                continue
+
+            declaration_invoiced = declaration.invoiced
+            declaration_not_yet_invoiced = declaration.not_yet_invoiced
+            # Exactly the confirmed SOs (state == 'sale') are included in `declaration.not_yet_invoiced`.
+            # The amount of `declaration.not_yet_invoiced` may change due to confirming or saving `order`.
+            #   * An unconfirmed order is being confirmed:
+            #     We have to add the declaration amount to `declaration.not_yet_invoiced`.
+            #   * A confirmed SO is being edited:
+            #     The field `declaration.not_yet_invoiced` will be updated when saving.
+            #     But we want to update the warning during the editing already (before saving).
+            #     We first have to remove the "old amount" from `declaration.not_yet_invoiced`
+            #     before adding the current amount.
+            if order.state == 'sale':
+                old_order_state = order._origin
+                declaration_not_yet_invoiced -= old_order_state.l10n_it_edi_doi_not_yet_invoiced
+            declaration_not_yet_invoiced += order.l10n_it_edi_doi_not_yet_invoiced
+
+            date = order.l10n_it_edi_doi_date
+            validity_warnings = declaration._get_validity_warnings(
+                order.company_id, order.partner_id.commercial_partner_id, order.currency_id, date, sales_order=True
+            )
```

Comment:
I know I'm being picky, but I like consistency (if you did not already know..).
So I would either pass the arguments directly, like you're doing here (which is my personal preference :blush:), or create a dedicated variable, like you're doing in `_compute_l10n_it_edi_doi_id()` (which is a bit unnecessary IMO).
Whatever you choose, I think it's better to be consistent :nerd_face: